tests/prague/eip7692_eof_v1/eip7069_extcall/test_address_space_extension.py

```diff
@@ -0,0 +1,226 @@
+        pytest.param(b"\xff" * 32, id="max_ase"),
+    ),
+)
+@pytest.mark.parametrize(
+    "target_account_type", ("empty", "EOA", "LegacyContract", "EOFContract"), ids=lambda x: x
+)
+@pytest.mark.parametrize(
+    "target_opcode",
+    (
+        Op.CALL,
+        Op.CALLCODE,
+        Op.STATICCALL,
+        Op.DELEGATECALL,
+        Op.EXTCALL,
+        Op.EXTDELEGATECALL,
+        Op.EXTSTATICCALL,
+    ),
+)
+def test_address_space_extension(
+    state_test: StateTestFiller,
+    target_address: bytes,
+    target_opcode: Op,
+    target_account_type: str,
+):
+    """
+    Test contacts with possibly extended address and fail if address is too large
+    """
+    env = Environment()
+
+    ase_address = len(target_address) > 20
+    stripped_address = target_address[-20:] if ase_address else target_address
+    if ase_address and target_address[0] == b"00":
+        raise ValueError("Test instrumentation requires target addresses trim leading zeros")
+
+    match target_opcode:
+        case Op.CALL | Op.CALLCODE:
+            call_suffix = [0, 0, 0, 0, 0]
+            ase_ready_opcode = False
+        case Op.DELEGATECALL | Op.STATICCALL:
+            call_suffix = [0, 0, 0, 0]
+            ase_ready_opcode = False
+        case Op.EXTCALL:
+            call_suffix = [0, 0, 0]
+            ase_ready_opcode = True
+        case Op.EXTDELEGATECALL | Op.EXTSTATICCALL:
+            call_suffix = [0, 0]
+            ase_ready_opcode = True
+        case _:
+            raise ValueError("Unexpected opcode ", target_opcode)
+
+    pre = {
+        TestAddress: Account(
+            balance=1000000000000000000000,
+            nonce=1,
+        ),
+        address_caller: Account(
+            code=(
+                Op.MSTORE(0, Op.PUSH32(target_address))
+                + Op.SSTORE(
+                    slot_top_level_call_status,
+                    Op.CALL(50000, address_target, 0, 0, 32, 0, 0),
+                )
+                + Op.STOP()
+            ),
+            nonce=1,
+        ),
+        address_target: Account(
+            code=Container(
+                sections=[
+                    Section.Code(
+                        code=Op.SSTORE(
+                            slot_target_call_status,
+                            target_opcode(Op.CALLDATALOAD(0), *call_suffix),
+                        )
+                        + Op.SSTORE(slot_target_returndata, Op.RETURNDATALOAD(0))
+                        + Op.STOP,
+                        code_inputs=0,
+                        code_outputs=NON_RETURNING_SECTION,
+                        max_stack_height=1 + len(call_suffix),
+                    )
+                ],
+            )
+            if ase_ready_opcode
+            else Op.SSTORE(
+                slot_target_call_status,
+                target_opcode(Op.GAS, Op.CALLDATALOAD(0), *call_suffix),
+            )
+            + Op.SSTORE(slot_target_returndata, Op.RETURNDATALOAD(0))
```

Comment:
```suggestion
            + Op.RETURNDATACOPY(0, 0, Op.RETURNDATASIZE)
            + Op.SSTORE(slot_target_returndata, Op.MLOAD(0))
```
Had to make this change locally to make some of the tests work, I'm not sure if `RETURNDATALOAD` is available for legacy contracts, and also want to make sure this is not the purpose of the test.

Reply:
RETURNDATALOAD is part of EOF, this particular test shouldn't be run in legacy.  evmone is out of spec for this test. It should work per spec - https://github.com/ethereum/EIPs/pull/8617 just got merged into the EIP.  @chfast @pdobacz

This is why I don't think we should add blocking github actions on tests targeting forks that haven't flipped any of the long-term test nets.  The spec could change and implementations could be implementing different versions of the EIP.

Reply:
I'm not 100% sure I'm following what's happening here but:
- evmone has RETURNDATALOAD implemented for EOF code only, like it should and like it is used in this test
- evmone doesn't just yet have the lax OoB for RETURNDATALOAD/COPY - [will be merged soon](https://github.com/ethereum/evmone/pull/909), causing the RETURNDATALOAD instruction to fail, however I think this in particular is slightly misleading, because:
- the change from LOAD to COPY seems to be covering up the real problem. How? I am guessing that the underlying `target_opcode` fails (see below why), and RETURNDATASIZE is 0, hence the change @marioevz made here changed semantics
- I am also guessing the underlying call fails because evmone doesn't yet implement ASE-ready EXT*CALL instructions - this is WIP right now

If I'm correct above, then what I think should happen is we should keep using RETURNDATALOAD and defer `ase_ready_opcode` testing until it is implemented in evmone. Additionally, since https://github.com/ethereum/EIPs/pull/8617 will soon become reality, we should make sure that the expected return data is not zero bytes, because this is what the RETURNDATA* instructions will give you even on accidental out-of-bounds read.

Reply:
I've removed the use of RETURNDATALOAD from these tests.  However they will return to a new test given the recent change to soft-zeroing in EOF, those need some explicit tests to validate correctness.

What remains for evmone are two checks:
* Revert on ASE addresses
* Revert on EXTDELEGATECALL 

While the delegate check may warrant an independent test, removing it from the matrix would leave a hole that is essentail to what is tested, whereas RETURNDATALOAD is arguably orthagonal.

Reply:
> defer ase_ready_opcode testing until it is implemented in evmone

I phrased this badly. I should have written: "defer testing ase_ready_opcode in evmone", not that tests wait on evmone, sorry about this. BTW This has just [been merged](https://github.com/ethereum/evmone/pull/915).

Re
>What remains for evmone are two checks:
As of today's master, both are implemented. EXTDELEGATECALL check has been for a while, was it failing to fail on such calls?

---

I've read the test again carefully and see the point about RETURNDATALOAD - you don't care if the zeros returned are due to noop RETURNDATACOPY (RETURNDATASIZE==0) or from zero-padding on RETURNDATALOAD. because on success we expect the caller address to be there - all clear now.